src/common/util.ts

```diff
@@ -0,0 +1,28 @@
+import { Result } from './evaluable'
+import { isNumber, isString } from './type-check'
+
+/**
+ * Convert a value to number if possible, otherwise return undefined
+ * @param value value to be converted to number
+ */
+export const toNumber = (value: Result): number | undefined => {
+  const isValueNumber = isNumber(value)
+  if (isValueNumber) {
+    return value as number
+  } else if (isString(value)) {
+    if (value.match(/^\d+\.\d+$/)) {
+      return parseFloat(value)
+    } else if (value.match(/^0$|^[1-9]\d*$/)) {
+      return parseInt(value)
+    }
+  }
+  return undefined
+}
+
+/**
+ * Convert a value to string if possible, otherwise return undefined
+ * @param value value to be converted to string
+ */
+export const toString = (value: Result): string | undefined => {
+  return value ? `${value}` : undefined
```

Comment:
should an empty string return `undefined`?

Reply:
I would say no, an empty string is a valid string

Reply:
I worded that too carefully, I believe that will return `undefined` for empty strings.

Reply:
I question why someone will need to cast a string to a string by using `.(String)` in the condition. Expectation is that the cast usage is solely for conversion from one type to another. I've update to support `0` though.

cc8306c

Reply:
It may simply a mistake, or invalid data context, so better to capture it.

Reply:
Definitely agree on `0 => '0'`, but I don't think we should reason our way out of fixing empty strings. Sometimes it's hard to reason about the path the code will take (today or in the future). I would not expect `toString` to return `undefined` for empty strings, and it could be a potential future source of bugs. However unlikely, I think it should be fixed. Maybe make the check `is nil` for null and undefined specifically instead?